### PR TITLE
PB-34214 local language support

### PR DESCRIPTION
**Pull request by michalmaciejewski91**

### CODE REVIEW

File: sdk/src/main/java/com/silanis/esl/sdk/builder/SignerBuilder.java [hunk 1]
```diff
@@ -44,6 +44,7 @@
     private String placeholderName = null;
     private List<AttachmentRequirement> attachments = new ArrayList<AttachmentRequirement>();
     private KnowledgeBasedAuthentication knowledgeBasedAuthentication;
+    private String localLanguage;
```
Comment (SeungChoi) [inline sdk/src/main/java/com/silanis/esl/sdk/builder/SignerBuilder.java]: I have got the answer to my question about local language features. The local language is a type of language code, and it's strange to add an extra property to the signer model for this. In other words, it seems correct to add local as the value of a language attribute that exists at the current signer.

By default, the language priority is:

user language> signer language> package language

I think that the local language should go in one language code, not in a separate attribute.

**Comment by SeungChoi** [comment]
I have got the answer to my question about local language features. The local language is a type of language code, and it's strange to add an extra property to the signer model for this. In other words, it seems correct to add local as the value of a language attribute that exists at the current signer.

By default, the language priority is:

user language> signer language> package language

I think that the local language should go in one language code, not in a separate attribute.

**Comment by michalmaciejewski91** [comment]
@SeungChoi , we made an agreement to validate the local language data value so it's possible to only provide "local". No "fr", "nl" etc. any more. In the future, we probably update this in some better way but this is what we have to release right now. For more information, check Sam's comment here: https://jira.onespan.com/browse/PB-35565 or contact him directly, 

Review and merge please. 
~Michał

File: sdk/src/main/java/com/silanis/esl/sdk/examples/LocalLanguageExample.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import static com.silanis.esl.sdk.builder.SignerBuilder.newSignerWithEmail;
+
+import com.silanis.esl.sdk.DocumentPackage;
+import com.silanis.esl.sdk.DocumentType;
+
+/**
+ * This example shows how to send a package with signer local language.
+ *
+ * Available localLanguages: en, fr, de, it, ru, zh-CN, zh-TW, es, pt, nl, da, el, ja, ko, local
+ */
```
Comment (SeungChoi) [inline sdk/src/main/java/com/silanis/esl/sdk/examples/LocalLanguageExample.java]: What does this mean? Does it mean you can use the above languages as a local language?

File: sdk/src/main/java/com/silanis/esl/sdk/internal/converter/SignerConverter.java [hunk 3]
```diff
@@ -139,6 +139,10 @@ private Signer newRegularSignerFromAPIRole() {
             signer.setLocked(true);
         }
 
+        if (apiRole.getData() != null && apiRole.getData().containsKey(Role.LOCAL_LANGUAGE_DATA_KEY)) {
+            signer.setLocalLanguage(apiRole.getData().get(Role.LOCAL_LANGUAGE_DATA_KEY).toString());
+        }
+
```
Comment (SeungChoi) [inline sdk/src/main/java/com/silanis/esl/sdk/internal/converter/SignerConverter.java]: As far as I know, we use the data from Role to set the local language. But do you need to do this? The SDK needs to handle the local language simply by passing data to the backend.

File: sdk/src/main/java/com/silanis/esl/sdk/internal/converter/SignerConverter.java [hunk 4]
```diff
@@ -173,6 +177,10 @@ private Signer newSignerPlaceholderFromAPIRole() {
...
```
Comment (SeungChoi) [inline sdk/src/main/java/com/silanis/esl/sdk/internal/converter/SignerConverter.java]: As far as I know, we use the data from Role to set the local language. But do you need to do this? The SDK needs to handle the local language simply by passing data to the backend.

File: tester/src/test/java/com/silanis/esl/sdk/examples/LocalLanguageExampleTest.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+import org.junit.Test;
+
+public class LocalLanguageExampleTest {
+
+    @Test
+    public void verifyResult() {
+        LocalLanguageExample example = new LocalLanguageExample();
+        example.run();
+    }
```
Comment (SeungChoi) [inline tester/src/test/java/com/silanis/esl/sdk/examples/LocalLanguageExampleTest.java]: What are you testing here? A test that doesn't do anything makes no sense. Please remove the test.

File: sdk/src/main/java/com/silanis/esl/sdk/Signer.java [hunk 6]
```diff
@@ -343,4 +344,12 @@ public KnowledgeBasedAuthentication getKnowledgeBasedAuthentication() {
...
         this.knowledgeBasedAuthentication = knowledgeBasedAuthentication;
     }
+
+    public String getLocalLanguage() {
+        return localLanguage;
+    }
+
+    public void setLocalLanguage(String localLanguage) {
+        this.localLanguage = localLanguage;
+    }
```
Comment (SeungChoi) [inline sdk/src/main/java/com/silanis/esl/sdk/Signer.java]: Do we need this local language field to use the local language? We set the local language as data in Role.
Why do we need duplicate settings?

File: sdk/src/main/java/com/silanis/esl/sdk/builder/SignerBuilder.java [hunk 7]
```diff
@@ -179,7 +180,7 @@ private Signer buildGroupSigner(){
         result.setMessage(message);
         result.setId(id);
         result.setAttachmentRequirements(attachments);
-
+        result.setLocalLanguage(localLanguage);
```
Comment (SeungChoi) [inline sdk/src/main/java/com/silanis/esl/sdk/builder/SignerBuilder.java]: Do we need this local language field to use the local language? We set the local language as data in Role.
Why do we need duplicate settings?

File: sdk/src/main/java/com/silanis/esl/sdk/builder/SignerBuilder.java [hunk 8]
```diff
@@ -216,7 +218,7 @@ private Signer buildRegularSigner(){
         result.setId(id);
         result.setAttachmentRequirements(attachments);
         result.setKnowledgeBasedAuthentication(knowledgeBasedAuthentication);
-
+        result.setLocalLanguage(localLanguage);
```
Comment (SeungChoi) [inline sdk/src/main/java/com/silanis/esl/sdk/builder/SignerBuilder.java]: Do we need this local language field to use the local language? We set the local language as data in Role.
Why do we need duplicate settings?

File: sdk/src/main/java/com/silanis/esl/sdk/builder/SignerBuilder.java [hunk 9]
```diff
@@ -377,6 +379,11 @@ public SignerBuilder withRoleId(String roleId) {
         return withCustomId(roleId);
     }
 
+    public SignerBuilder withLocalLanguage() {
+        this.localLanguage = "local";
+        return this;
+    }
+
```
Comment (SeungChoi) [inline sdk/src/main/java/com/silanis/esl/sdk/builder/SignerBuilder.java]: Do we need this local language field to use the local language? We set the local language as data in Role.
Why do we need duplicate settings?

File: sdk/src/main/java/com/silanis/esl/sdk/examples/LocalLanguageExample.java (hunk 2 above)
Comment (michalmaciejewski91) [reply sdk/src/main/java/com/silanis/esl/sdk/examples/LocalLanguageExample.java]: This was the old comment from the solution which is not valid any more. I've removed it now.

File: tester/src/test/java/com/silanis/esl/sdk/examples/LocalLanguageExampleTest.java (hunk 5 above)
Comment (michalmaciejewski91) [reply tester/src/test/java/com/silanis/esl/sdk/examples/LocalLanguageExampleTest.java]: test deleted

**Comment by michalmaciejewski91** [comment]
@SeungChoi , answering to all of the unresolved conversations and the questions about the localLanguage attribute on both, Signer and Role objects I can say that for the user / integrator local language is on the Signer lvl. Sender sets the local language for the Signer and he's not interested what is happening on the background. That's fine and that's how it supposed to be. The reason we have to duplicate this information in Role Data is that this is how it's implemented on the backend and was discussed and explained many times in the past already. If we would have to leave only 1 localLanguage, I mean, only on Signer OR only on Role data, we would have to remove it from Signer but how should Builder looks like then? How would you write an example and force user to set this on the Role if there is no RoleBuilder in the SDK? It does not make sense to create a new one just for this. Role is created ALWAYS based on the Signer configuration and it's 1 to 1 all the time.

**Comment by SeungChoi** [comment]
@michalmaciejewski91 I understood that you had to add that attribute to the signer because we didn't use the concept of Role in the SDK.
But what I wanted to say is why you specify the local language in the Data property of Role, not the signer. If we set a local language value in the signer's data, this would not be necessary.
In fact, the signer language is assigned to the signer, not to Role. This is again reflected in the language in the ESL_SIGNER_SUMMARY table.
If you specify a local language for Role as it is now, there is a possibility of importing the wrong language when retrieving language information from ESL_SIGNER_SUMMARY.
If there is no reason to specify a local language for Role, why not assign it to a signer?

**Comment by michalmaciejewski91** [comment]
@SeungChoi , there is no signer's data attribute in Groovy layer.

**Pull request merged**